### DISCUSSION TITLE
Alternate location for Templates for Folder Templates

### DISCUSSION THREAD

**Discussion by djdanner** [SilentVoid13/Templater#1264] [Ideas & Feature Requests]
For Folder Templates, instead of specifying the location of a Folder's Template in a Templater Setting, could a Note named "Template.md" simply be placed in the Folder?  Then any Note created in that Folder, or any of its Subfolders, would use the Template in that Folder.

If there were multiple Template.md files in a Folder Hierarchy, you would stop looking as soon as you found a Template when going-up the Parent Chain.

This could be done as an alternative to the existing method so that both methods are supported.

**Comment by Zachatoo** [comment]
Templater is so extensible, you can do this today!

Here's a Templater script you can use to accomplish this. Set this as your only folder template at the root folder (`/`).

```js
<%*
// Search for "Template.md" note recursively, starting at the current folder and ending at the root folder
let folder;
let folderTemplate;
do {
  if (!folder) {
    folder = app.vault.getAbstractFileByPath(tp.file.folder(true));
  } else {
    folder = folder.parent;
  }
  folderTemplate = folder.children.find(x => x.name === "Template.md");
} while (!folderTemplate && folder.path !== "/")
// Apply folder template (if applicable)
if (folderTemplate) {
  tR += await tp.file.include(folderTemplate);
}
-%>
```

**Reply by djdanner** [reply]
That is awesome!   Thank you!

I just made the following change so that it will look for any file that ends with "_Template.md".  This way I can use a unique filename for each Template File so that I don't end-up with hundreds of files named "Template.md".

     folderTemplate = folder.children.find(x => x.name.endsWith("_Template.md"));

Thank you again for your help.  This will greatly simplify using Folder Templates!

**Reply by djdanner** [reply]
Now that I've thought about how this works a little more, why does Templater look in the Root Folder `\`, when I'm creating a Note in some other Folder?

**Reply by Zachatoo** [reply]
It will eventually look in the root folder, if as it's searching up the folder chain it can't find one anywhere else.

If it's not working as you're expecting, you can add in a `debugger;` line in this template and step through the logic as it's happening. Was very useful for me when I was testing.

**Reply by djdanner** [reply]
Thanks, I understand how the script works to find a Template, regardless of where it is in a new Note's Parent-chain.  What I don't understand is how the Template with the Script in it (at the Root, /) got called/executed.   Are you saying that the default Templater operation is to walk the Note's Parent-chain looking for a Template?

**Reply by Zachatoo** [reply]
Ah, I see. Yes, but there's some nuance if you have multiple folder templates.

The default is to look at each folder template, from top to bottom. For each folder template, it will then check if the newly created file is somewhere in that folder, including if it's deeply nested (walking the parent chain). If it is, execute that template and stop looking at later folder templates.

So, if you want a template for a deeply nested folder template to override a template for a folder template that isn't as deeply nested or not nested at all, then you should put the more specific one above the more general one.

**Reply by djdanner** [reply]
I'm still confused : )  What I don't understand, is, when I click "Create a New Note", does Obsidian jump straight to the Template with the Script in it?   If so, how does it know to do that?  If not, what's the sequence that gets it there?  Is there any way we can communicate via a higher bandwidth channel (eg, voice)?

**Reply by Zachatoo** [reply]
Here's a video of me explaining how it works. Or at least how I was expecting it to work, one part of the video didn't make sense with how I thought it worked. But hopefully it's helpful.

[2023-12-22_13-23-55.mp4.zip](https://github.com/SilentVoid13/Templater/files/13756553/2023-12-22_13-23-55.mp4.zip)

Sorry it's a zip file, GitHub won't let me upload files over 10MB.

**Reply by djdanner** [reply]
Awesome!  Thanks!!!  If a picture is worth a thousand words, then a video is probably worth 10,000!   That was very helpful to see that I think we're talking about two different things.   If I can, I'll record a video to reply.  What did you use to record that?  Thx

**Reply by djdanner** [reply]
I just found this video that says that I should be able to use XBOX Game Bar to record it:
https://youtu.be/pRSLYvqU-tM

**Reply by djdanner** [reply]
Hey Zach, thanks again for that video.  It was a big help.  My reply video is in 2 parts, because 1 file was too big.
[Folder-Templates-PART-1.zip](https://github.com/SilentVoid13/Templater/files/13762189/Folder-Templates-PART-1.zip)

**Reply by djdanner** [reply]
Part 2:
[Folder-Templates-PART-2.zip](https://github.com/SilentVoid13/Templater/files/13762192/Folder-Templates-PART-2.zip)

**Reply by Zachatoo** [reply]
Watched the videos and it seems like towards the end of the second video we got on the same page. You're right, because your only folder template was setup to trigger on the root (`/`), then Templater will always use that root template for every new note creation, which is this script we wrote to have this new behavior of looking for a `_Template.md` file up the tree.

I do want to make one thing clear. If, for example, you only had one folder template but it was setup on a different folder than the root, say the `Folder-1` folder, then this script that we wrote would only take affect for files created in that folder or it's children folders. Files created in other folders that are children of the root or files created in the root folder itself would not be triggered by the folder template and therefore would not use our custom logic.

I'll make a post about this script (my variant and yours) to promote this idea. I have my own custom script that I use that I like that also modifies the default behavior of how folder templates work that I've promoted that a few people like, documented [here](https://zachyoung.dev/posts/folder-templates-with-quick-switcher). I'll do the same for this since I think this is another great way to manage templates (assuming you're okay with me posting your script alongside mine)!

**Reply by djdanner** [reply]
Great.  Yeah, the way I'm using it is that I have an _Template file in several folders and then I have a Default _Template file at the Root.  As I mentioned in the video, the only piece I was confused about was how Templater knew to go the Root to find the script when you were creating a Note in a subfolder.

Feel free to share any of this thread, and no need to credit me, but thanks for asking : )   Please send me a link if you do make a post for this topic, thx.  And I'll check-out your other referenced post.  

Thanks again for all your help!
